### ISSUE TITLE
feat: rewrite using macros

### ISSUE THREAD

**Issue by felangel** [felangel/equatable#175]
With [macros](https://dart.dev/language/macros) on the horizon, it will be possible to implement equatable using macros which will both improve the performance of equality comparisons and significantly simplify the usage.

**Before**
```dart
class Person extends Equatable {
  const Person({required this.name});
  final String name;

  @override
  List<Object> get props => [name];
}
```

**After**
```dart
@Equatable()
class Person {
  const Person({required this.name});
  final String name;
}
```

See also https://github.com/felangel/data_class which will be built on top of `package:equatable`.

**Comment by tienlbhoc** [comment]
This is a good news. But can you add more features:
Current:
- copyWith : seem enough, maybe change to copy (like jetpack compose to shorter)
- operator== and hashCode : seem need config like Equatable because a long response from server can check only by some fields.
- toString : I always change to string because some field need protect and list only show length so need custom option
- const : seem ok

+ Need option convert class to json and json to class and need json key map example:
@Data()
class Person {

  @key("full-name")
  final String fullName;
}

**Comment by felangel** [comment]
Equatable is meant to just be for equality comparisons. package:data_class will implement copyWith, toString, and equality. For json serialization check out package:json and the JsonCodable macro

**Comment by rainyl** [comment]
Great, but it will be better if we can specify the properties to compare like before, sometimes we just need several props to compare.

**Comment by felangel** [comment]
> Great, but it will be better if we can specify the properties to compare like before, sometimes we just need several props to compare.

In those cases you should probably just write your own operator== override since the goal of the package is to provide value based equality comparisons (which means taking all fields into account).

**Comment by rainyl** [comment]
> > Great, but it will be better if we can specify the properties to compare like before, sometimes we just need several props to compare.
> 
> In those cases you should probably just write your own operator== override since the goal of the package is to provide value based equality comparisons (which means taking all fields into account).

Allright, thx.

**Comment by Manishmg3994** [comment]
Subject: Hashing Approach for `equatable_macro`

Hi @felangel, @rrousselGit, @slightfoot

I have created a Dart package with macros similar to `equatable`, called [equatable_macro](https://github.com/antinna/equatable_macro). However, I am facing issues with the hashing implementation. I would appreciate your feedback on whether this approach for hashing is appropriate or if I should consider an alternative method.

### Current Hashing Implementation

#### Method 1: `hashAll`

```dart
int hashAll(Iterable<Object?>? fields, {Type? clazzIdentifier}) {
  // Use Object.hashAll to combine the class identifier's hash code with the hash codes of the fields
  return Object.hashAll([clazzIdentifier, ...?fields]);
} 
```
## Method 2: Class-specific Hashing
```dart
// Here `MacroEquatableExample` is the class identifier name and other fields like `id` and `name` are properties
augment prefix0.int get hashCode => prefix0.Object.hashAll([MacroEquatableExample, id, name]);
```
## Questions
- Is the approach using Object.hashAll with a class identifier and props good for hashing?
- Should I consider another way for hashing in macros?
I would greatly appreciate any guidance or suggestions you can provide.

Thank you!

**Comment by vadrian89** [comment]
![image](https://github.com/user-attachments/assets/c1e32915-7e69-4f89-a752-d29ab2e232e5)

Hello Felix, i stumbled upon this issue and I noticed both you and Remi Rousselet seem to be building the same type of package. Wouldn't be more beneficial if you and him join up?

**Comment by felangel** [comment]
> ![image](https://private-user-images.githubusercontent.com/8177798/357104528-c1e32915-7e69-4f89-a752-d29ab2e232e5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkwNDU2NzMsIm5iZiI6MTcyOTA0NTM3MywicGF0aCI6Ii84MTc3Nzk4LzM1NzEwNDUyOC1jMWUzMjkxNS03ZTY5LTRmODktYTc1Mi1kMjlhYjJlMjMyZTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTZUMDIyMjUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmNiYzhmYmVlMDliMTM2YjJiNmFmMmU2YTJlODMxMzE4M2I5ZDNjMDgxNzljYWI3ODZjNTQyYjQzMmNjZjM2NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.fLcnKN0Q2i9lUe1MPi0Gjur2jRmmNRryF_oGSJruH5w)
> 
> Hello Felix, i stumbled upon this issue and I noticed both you and Remi Rousselet seem to be building the same type of package. Wouldn't be more beneficial if you and him join up?

Remi expressed his interest in using `package:dataclass` to migrate `package:freezed` which has a different scope, goals, API etc.

**Comment by felangel** [comment]
This is now published as a dev release in [v3.0.0-dev.0](https://pub.dev/packages/equatable/versions/3.0.0-dev.0) 🎉
